bundles/io/org.eclipse.smarthome.io.transport.bluetooth/src/main/java/org/eclipse/smarthome/io/transport/bluetooth/BluetoothGatt.java

```diff
@@ -0,0 +1,460 @@
+                            logger.debug(": Sending CHARACTERISTIC_READ {}", sentMessage.getCharacteristic().getUuid());
+                            sendReadCharacteristic(sentMessage.getCharacteristic());
+                            break;
+                        case CHARACTERISTIC_WRITE:
+                            logger.debug(": Sending CHARACTERISTIC_WRITE {}",
+                                    sentMessage.getCharacteristic().getUuid());
+                            sendWriteCharacteristic(sentMessage.getCharacteristic());
+                            break;
+                        case DESCRIPTOR_READ:
+                            logger.debug(": Sending DESCRIPTOR_READ {}", sentMessage.getDescriptor().getUuid());
+                            sendReadDescriptor(sentMessage.getDescriptor());
+                            break;
+                        case DESCRIPTOR_WRITE:
+                            logger.debug("Sending DESCRIPTOR_WRITE {}", sentMessage.getDescriptor().getUuid());
+                            sendWriteDescriptor(sentMessage.getDescriptor());
+                            break;
+                        default:
+                            logger.debug("Sending unknown {}", sentMessage.getMessageType());
+                            break;
+                    }
+                }
+            } catch (Exception e) {
+                logger.error("Exception during Send thread", e);
+            }
+
+            logger.debug("Exiting transmit thread");
+        }
+    };
+
+    protected void sendReadCharacteristic(BluetoothGattCharacteristic characteristic) {
+    }
+
+    protected void sendWriteCharacteristic(BluetoothGattCharacteristic characteristic) {
+    }
+
+    protected void sendReadDescriptor(BluetoothGattDescriptor descriptor) {
+    }
+
+    protected void sendWriteDescriptor(BluetoothGattDescriptor descriptor) {
+    }
+
+    /**
+     * This class implements a synchronisation mechanism to wait until we are free to send.
+     *
+     * The transmit queue calls waitForSend which will block until we're connected, there's a message
+     * to send, and there's no outstanding messages awaiting responses.
+     *
+     */
+    private class TransmitNotify {
+        private Object myMonitorObject = new Object();
+        private boolean isConnected = false;
+        private boolean isFreeToSend = false;
+
+        public void waitForSend() {
+            synchronized (myMonitorObject) {
+                logger.debug("waitForSend start: connected={}, free={}, queue={}", isConnected, isFreeToSend,
+                        sendQueue.size());
+                while (isConnected == false || isFreeToSend == false || sendQueue.size() == 0) {
+                    logger.debug("waitForSend waiting: connected={}, free={}, queue={}", isConnected, isFreeToSend,
+                            sendQueue.size());
+                    try {
+                        myMonitorObject.wait();
```

Comment:
I'm not so sure. When adding to the queue, we trigger a connect - and it's the result of the connect that then triggers the queue - through `setConnectedState`. If we started sending just because we added a message to the queue, then we'd have problems if not connected. In the case of BlueZ, this is not done in this class, but in the BluezBluetoothGatt class.
Have you actually seen an issue with this, or is it just from code review? If you are having a problem, then I'll look further, but I've not noticed this locking up.